### PR TITLE
docs: add resource types reference page

### LINKED ISSUE #22298: PVC resource requirements hard to find
**This is a Bug Report**

**Problem:**
I was looking for the keys to specify for the resource requirements in a PersistentVolumeClaim via the [API reference][api-ref-pvc]. After following the link to the [ResourceRequirements][resource-requirements] docs, the `.requests` key points to the [concepts][concepts] page for managing container resources. It makes no mention of Persistent Volume Claims or the fact that they take the key `.storage` key in their `.requests` field.

The way to find out the key is going to the [PersistentVolumeClaim][pvc-concepts] page and looking at one of the examples 
**Proposed Solution:**
Ideally, the reference page for [ResourceRequirements][resource-requirements] reference page includes all the keys that would satisfy a resource requirement 

e.g. from the concepts page:  
-   `.requests.cpu` (PodSpec)
-   `.requests.memory` (PodSpec)
-   `.requests.hugepages-<size>` (PodSpec)
with new PVC info
- `.requests.storage` (PersistentVolumeClaimSpec)

If this is not desired, updating the PVC concepts page to include a section on persistent volume claim resource requirements and updating the API reference page to link to the PVC page as well 
**Page to Update:**
https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#resourcerequirements-v1-core

https://kubernetes.io/docs/concepts/storage/persistent-volumes/

https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-types

`v1.18`

[api-ref-pvc]: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#persistentvolumeclaimspec-v1-core
[resource-requirements]: https://kubernetes.io/docs/reference/generated/kubernetes-api/v1.18/#resourcerequirements-v1-core
[concepts]: https://kubernetes.io/docs/concepts/configuration/manage-compute-resources-container/
[pvc-concepts]: https://kubernetes.io/docs/concepts/storage/persistent-volumes/#persistent-volume-claim-requesting-a-raw-block-volume

### PR DESCRIPTION
**Pull request by kreempuff**
Going off of what was discussed in #22298, I created a new
page to describe the different resource types available in Kubernetes.

I only included the resource names that had the `v1.ResourceName` type excluding
`ephemeral-storage` because it was labeled as alpha.

I wasn't sure how the `v1.ResourceName` towards the [bottom](https://github.com/kubernetes/kubernetes/blob/b2ecd1b3a3192fbbe2b9e348e095326f51dc43dd/staging/src/k8s.io/api/core/v1/types.go#L5568)
of the files were used so I did not include them.

I'm looking to see if I'm headed in the right direction. 

I'm still going through the style [guide](https://kubernetes.io/docs/contribute/style/style-guide/).

Closes #22298

### CODE REVIEW

**Review by Arhell — COMMENTED** [review]
Hi, you can check, the mail on github and cla must match.

**Comment by kreempuff** [comment]
> Hi, you can check, the mail on github and cla must match.

I'm having a bit of trouble requesting the sign email

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:3]:
```suggestion
---
title: Resource Types
content_type: reference
---
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:7]:
```suggestion
This page provides information on the different resource types you can request, built into Kubernetes.
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:11]: Hi. There is also `hugepages-*` - see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-requests-and-limits-of-pod-and-container

and, as a beta feature, `ephemeral-storage` - see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#local-ephemeral-storage and https://kubernetes.io/docs/concepts/policy/resource-quotas/#storage-resource-quota

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:14]: I recommend explaining “CPU” in more detail; see https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu for some wording you could migrate to this page.

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [reply content/en/docs/reference/resource-types/_index.md:7]: Maybe this page should also mention that [extended resources](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources) exist?

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:26]:
```suggestion
You can use the `storage` resource in a PersistentVolumeClaim.
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:21]:
```suggestion
You can request or limit the `memory` resource for Pods and their containers, and any for object that embeds a Pod template.
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:16]:
```suggestion
You can request or limit the `cpu` resource for Pods and their containers, and any for object that embeds a [Pod template](https://kubernetes.io/docs/concepts/workloads/pods/#pod-templates).
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:29]: nit:
```suggestion
### CPU and memory
```
as [headings](https://kubernetes.io/docs/contribute/style/style-guide/#headings) are usually sentence case.

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:12]: I recommend telling the reader that resources are relevant to [scheduling](https://kubernetes.io/docs/concepts/scheduling-eviction/) as well as for enforcing limits.

**Review by sftim — CHANGES_REQUESTED** [review]
Hi @kreempuff 

Thanks for this PR. I'd like to see some changes because Kubernetes actually supports more than 3 resource types.

I have made some suggestions inline. Do these make sense? I'll be happy to explain my thinking.

File: content/en/docs/reference/resource-types/_index.md
Comment (kreempuff) [reply content/en/docs/reference/resource-types/_index.md:3]: Where can I find the list of `content_type`s?

File: content/en/docs/reference/resource-types/_index.md
Comment (kreempuff) [reply content/en/docs/reference/resource-types/_index.md:7]: I agree!

File: content/en/docs/reference/resource-types/_index.md
Comment (kreempuff) [reply content/en/docs/reference/resource-types/_index.md:11]: Copy, I'll add these!

File: content/en/docs/reference/resource-types/_index.md
Comment (kreempuff) [reply content/en/docs/reference/resource-types/_index.md:12]: Sounds good!

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:17]: We always use relative hyperlinks:
```suggestion
You can request or limit the `cpu` resource for Pods and their containers, and any for object that embeds a [Pod template](/docs/concepts/workloads/pods/#pod-templates).
```

**Review by sftim — CHANGES_REQUESTED** [review]
One required tweak, then I reckon this is good to merge.

**Comment by sftim** [comment]
Hi @kreempuff 

Would you be willing to revise this PR as previously suggested? I'd love to see a tweaked version of this merge into the repo.

**Comment by kreempuff** [comment]
> Hi @kreempuff
> 
> Would you be willing to revise this PR as previously suggested? I'd love to see a tweaked version of this merge into the repo.

Yes, of course! EOY was just a busy time for me

**Comment by kreempuff** [comment]
@sftim updated based on your feedback. I did not find any internal references to huge pages to link to like I did with the other resource types

Also, thanks for your patience on this PR!

**Comment by kreempuff** [comment]
@sftim nudging

File: content/en/docs/reference/resource-types/_index.md
Comment (kbhawkey) [inline content/en/docs/reference/resource-types/_index.md:19]: nit: Would you add a newline before the CPU heading? Thanks!

**Comment by kbhawkey** [comment]
Page preview:
https://deploy-preview-25177--kubernetes-io-master-staging.netlify.app/docs/reference/resource-types/

**Comment by kbhawkey** [comment]
Hi @kreempuff .
Thanks for this contribution!
/lgtm

**Comment by tengqm** [comment]
Frankly speaking, I failed to see the value of this reference page. Almost all of the contents provided in this reference page can be found in the concepts section and the related tasks section.

**Comment by tengqm** [comment]
What's more, "resource type" as a reference page may be misleading. In the k8s context, a resource may refer to the CPU, memory, storage or other custom resources that can be requested by workloads; it may refer to the API resource such as 'Pod', 'Deployment' etc.

**Comment by kreempuff** [comment]
> Frankly speaking, I failed to see the value of this reference page. Almost all of the contents provided in this reference page can be found in the concepts section and the related tasks section.

When I raised the issue I was having trouble finding out how to specify resource requests for things other than CPU and memory. The documentation doesn't succinctly and directly tell me how to _specify_ certain resources. I think understanding the concept of the resource and completing a certain task including a resource is a different concern that detailing what values they can take on. That's how I view this new page. 

I think this sentence from the issue highlights the problem:

> The API reference doesn't have awareness of all the kinds of resources that you can request (or set a limit for).

If there's a way to generate this information as part of the API reference that would be better. I'm just not sure what that would entail. 

> What's more, "resource type" as a reference page may be misleading. In the k8s context, a resource may refer to the CPU, memory, storage, or other custom resources that can be requested by workloads; it may refer to the API resource such as 'Pod', 'Deployment' etc.

Would **_Resource requests and limits_** be a better title?

@tengqm

File: content/en/docs/reference/resource-types/_index.md
Comment (kreempuff) [inline content/en/docs/reference/resource-types/_index.md:19]:
```suggestion

## CPU 
```

Comment (kreempuff) [reply content/en/docs/reference/resource-types/_index.md:20]: Done!

**Comment by tengqm** [comment]
@kreempuff instead of adding a new reference, is it possible to improve the "https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/" page?

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:10]: This is a short overview IMO. Best to merge it with the paragraphs that follow. That way “CPU” is the first part of the page body.

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:57]: The correct spelling is “Ephemeral”.

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:60]:
```suggestion
long-term guarantee of durability. Ephemeral storage is in beta.
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:20]: How about making these be a 3rd level heading, under a 2nd level heading “resource types”.

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:67]: We should add a heading for custom (extended) resource types - ie, https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#extended-resources

**Review by sftim — DISMISSED** [review]
Hi. This is going to need a few changes before it is ready to merge.

I would like to see a list of resource types in the reference section. The reference section is the right place to put exhaustive lists of different kinds of things that are built-in to Kubernetes (like built-in resource types).

File: content/en/docs/reference/resource-types/_index.md
Comment (kbhawkey) [inline content/en/docs/reference/resource-types/_index.md:2]: I think the title of the page (and the directory) should be changed. I was initially confused
by the use of the "resource".

**Comment by kbhawkey** [comment]
Hold for review
/hold

**Comment by sftim** [comment]
It's tricky. The English word “resource” has lots of meanings. There's possibly a separate issue about explaining the different meanings in the glossary.

The things that you refer to in requests and limits _are_ resources, even if the things you manage via the cluster's HTTP API _also are_ resources.

**Comment by sftim** [comment]
Also relevant: #20074

File: content/en/docs/reference/resource-types/_index.md
Comment (celestehorgan) [reply content/en/docs/reference/resource-types/_index.md:2]: Would "PersistentVolumeClaim Resource Types" clarify it further?

**Comment by celestehorgan** [comment]
@kreempuff, I really love this contribution and I'm keen to see it merged, but we need to address the issues raised by other reviewers to do so.

Do you intend to do this? If not, I'd love to have another contributor jump into this PR and take it over the line so we can merge in this great content :)

**Comment by kreempuff** [comment]
> @kreempuff, I really love this contribution and I'm keen to see it merged, but we need to address the issues raised by other reviewers to do so.
> 
> Do you intend to do this? If not, I'd love to have another contributor jump into this PR and take it over the line so we can merge in this great content :)

@celestehorgan Ahh apologies. I just didn't realize how active contributing to open source is! 😓 I do intend to get this done. I'll address the comments this week!

File: content/en/docs/reference/resource-types/_index.md
Comment (kreempuff) [reply content/en/docs/reference/resource-types/_index.md:21]: I believe this was resolved below. Feel free to re-add this comment if you feel it wasn't addressed!

**Comment by jimangel** [comment]
> The English word “resource” has lots of meanings.

I agree strongly with @sftim and other's comments around the title. I don't have a better alternative at the moment. Considering this PR has been open since November 2020, I would like to see it merged and iterated on top of if we cannot find an alternative title. Thoughts?

/cc @kbhawkey

**Comment by kbhawkey** [comment]
Page preview:
https://deploy-preview-25177--kubernetes-io-main-staging.netlify.app/docs/reference/resource-types/

**Comment by kbhawkey** [comment]
👍

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:119]:
```suggestion
### Ephemeral storage {#example-ephemeral-storage}
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:26]: (nit)
```suggestion
The container runtime ensures that processes in a container do not exceed their CPU allocation.

See [Meaning of CPU](/docs/concepts/configuration/manage-resources-containers/#meaning-of-cpu)
for details of CPU units in Kubernetes.
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:34]: How about:
```suggestion
If a container or Pod tries to use more memory than the configured limit, the
container runtime and / or host operating system kernel typically intervenes by
stopping a process in that Pod. That mechanism is called
“out of memory kill”, and is different from how CPU resource limits are handled

(CPU limits are enforced using time-slicing, whereas memory limits are enforced
by stopping containers that try to violate them).
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:36]: nit:
```suggestion
See [Meaning of Memory](/docs/concepts/configuration/manage-resources-containers/#meaning-of-memory)
for more details on what memory means in Kubernetes.
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:41]: (nit)
```suggestion
You can use the `storage` resource in a [PersistentVolumeClaim](/docs/concepts/storage/persistent-volumes/#persistentvolumeclaims).
```
or, even better:
```suggestion
You can use the `storage` resource in a
{{< glossary_tooltip text="PersistentVolumeClaim" term_id="persistent-volume-claim" >}}.
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:105]:
```suggestion
### Huge pages {#example-hugepages}
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:92]:
```suggestion
### Storage {#example-storage}
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:75]:
```suggestion
### CPU and memory {#example-cpu-memory}
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:48]: Can we assume the reader is using a supported version of Kubernetes?
```suggestion
Huge pages are a Linux-specific feature where the node 
kernel allocates blocks of memory that are much larger than the 
default page size. You can set requests or limits for particular sizes
of huge page.
```

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:54]: for example, `hugepages-2Mi`.
That `2Mi` specifies the size of the pages; in this case, 2 MiB.
The second part is the value of allocatable pages. For example, `hugepages-2Mi: 100Mi`
allows using up to 50 huge pages, 2 MiB each, for a total of 100 MiB.
See the [example](#example-hugepages) to see this in context.

File: content/en/docs/reference/resource-types/_index.md
Comment (sftim) [inline content/en/docs/reference/resource-types/_index.md:2]:
```suggestion
title: Kubernetes Resource Management Types
```
?

**Review by sftim — COMMENTED** [review]
Some more suggestions. I'm happy to see this merge with the original proposed title “Resource Types” if that feels like a better fit that my suggestion, “Kubernetes Resource Management Types”

**Comment by shannonxtreme** [comment]
FWIW, [Managing Resources for Containers](https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-types) distinguishes between `compute resources` like CPU, memory and hugepages, and `API resources`.

File: content/en/docs/reference/resource-types/_index.md
Comment (tengqm) [inline content/en/docs/reference/resource-types/_index.md:41]: The verb 'use' is misleading.

File: content/en/docs/reference/resource-types/_index.md
Comment (tengqm) [inline content/en/docs/reference/resource-types/_index.md:52]: Referring this as 'a combination' is confusing.
The paragraphs above are an exact copy of https://kubernetes.io/docs/concepts/configuration/manage-resources-containers/#resource-types.

**Review by tengqm — COMMENTED** [review]
Again, I'd be unhappy if we add a new page like this to the website because 1) it adds no value to the existing content; 2) it creates duplicated contents, which will be a headache for consistency maintenance.

/hold

File: content/en/docs/reference/resource-types/_index.md
Comment (kreempuff) [reply content/en/docs/reference/resource-types/_index.md:48]: I think that's fair. I just like to thing of a newcomer that may be following old tutorials and they fall into some unclear errors

File: content/en/docs/reference/resource-types/_index.md
Comment (kreempuff) [reply content/en/docs/reference/resource-types/_index.md:14]: I realize it's a duplication but the value it adds imo is the locality to the references and examples. The original page is an in-depth explanation of the resources

Happy to here an alternative. Should I just remove the overview or reword it?

**Comment by kbhawkey** [comment]
👋 @kreempuff . Thanks for contributing.
I suggest that we close this PR. How could you present the resource content as a reference page (and not repeat the content from the concepts section). Feel free to reopen this pull request, if you would like to continue to work on these changes.
/close

**Comment by sftim** [comment]
I like the idea of a reference page that is authoritative, and keeping the equivalent wording in the concept page shorter.

If we can justify it we can make a reference page that is reusable elsewhere via a shortcode. I don't think that effort would pay off though.

**Comment by kbhawkey** [comment]
> 
> 
> I like the idea of a reference page that is authoritative, and keeping the equivalent wording in the concept page shorter.
> 
> If we can justify it we can make a reference page that is reusable elsewhere via a shortcode. I don't think that effort would pay off though.

This makes sense. I felt that after a year, this pull request should either merge or close. There are remaining questions about how to make this content different than the concepts section.

**Pull request closed without merging**
